### ISSUE TITLE
Use without sublevel

### ISSUE THREAD

**Issue by tnguyen14** [dominictarr/level-search#9]
Currently I have a leveldb created without using sublevel. Is it possible to use level-search on that?

**Comment by dominictarr** [comment]
no, this needs sublevel, but you can add sublevel just for this.

``` js
LevelSearch(SubLevel(db), 'search')
```

however, you are now responsible for making sure that the other stuff you add does not collide with sublevel's keyspace.

**Comment by tnguyen14** [comment]
Thanks @dominictarr. I thought so too, but for some reason I wasn't able to perform the search. Perhaps I am using level-search incorrectly?

Here's the code

``` js
var index = search(sub(db), 'transactions');

db.createReadStream({
    gte: 'transaction!',
    limit: 1
}).on('data', console.log);

//{ key: 'transaction!daily!1418839200000',
//  value:
//  { amount: 1000,
//     date: '2014-12-17T18:00:00.000Z',
//     description: 'T card',
//     merchant: 'T',
//    status: 'POSTED',
//     category: 'default' } }

index.createSearchStream(['status', 'POSTED'])
    .on('data', console.log);
// no output
```

About collision with sublevel keyspace - the sublevel page mentions that "Sublevel prefixes each subsection", but how does it do that exactly? 

Right now, I'm prefixing my db entries with namespace such as `transaction!`. If there's a `db.sublevel('transaction')`, would that have the same prefix?

**Comment by dominictarr** [comment]
there are several different prefixing strategies, by default it uses this:
https://github.com/dominictarr/level-sublevel/blob/master/codec/index.js

prefixes with `\x00` (null) and separates with `\xffff` so you'll be alright.

did you add `level-search` _before_ you added the data? because level-search creates indexes on write,
so you have to add it before writing anything.

which is realize is actually pretty awkward, and is why I have a new project https://github.com/flumedb/flumedb that adopts a very different architecture for a modular database, which solves that problem!

**Comment by tnguyen14** [comment]
@dominictarr awesome. thanks.

No, the data was not added before `level-search`. In that case, should I re-index level search somehow? That should just be a one-time operation?

I'll take a look at flumedb. Does that build on top of leveldb?

**Comment by dominictarr** [comment]
flumedb builds _beside_ leveldb. It's basically built around a log - where the data is stored - and then many views which are generated from the data in the log. Level is pretty good for an index/view, but not great at being a log.

**Comment by dominictarr** [comment]
If you just write the data back into the the database, you'll get the indexes generated again.

**Comment by tnguyen14** [comment]
flumedb sounds great. what kind of value can be stored with flume? Im assuming anything that can be serialized through JSON like level?

what do you mean when you say

> If you just write the data back into the the database

Would that not create duplicate entries?

**Comment by dominictarr** [comment]
well, it's still very new, but any type of value that can serialize to a buffer.

I meant if you do `db.createReadStream().pipe(writeStream(db))`

**Comment by tnguyen14** [comment]
What is `writeStream` in this case?

**Comment by dominictarr** [comment]
say,
https://www.npmjs.com/package/level-write-stream

originally level had a built in write stream but we took it out because there where too many ways to implement it

**Comment by tnguyen14** [comment]
I see. Thanks for the context. So my understanding is that that would not create duplicate entries, because it's `put`-ing with the same key and value, but would still trigger the search index? And this only needs to be done once?

**Comment by dominictarr** [comment]
yeah, except then I realized that it needs to call put where the sublevel can intercept it, which would duplicate the values, or better, copy everything currently in there to another db that uses sublevel.

**Comment by tnguyen14** [comment]
So the search index needs to be at the same sublevel as the db in order for it to work?

**Comment by dominictarr** [comment]
yeah, basically:

```
var db = sublevel(level(path))
search(db, 'search')
db.put(...)
```
